src/Pos/Util.hs

```diff
@@ -0,0 +1,377 @@
+
+mconcatPair :: (Monoid a, Monoid b) => [(a, b)] -> (a, b)
+mconcatPair = foldr mappendPair (mempty, mempty)
+
+-- | Concatenates two url part using regular slash '/'.
+-- E.g. @"./dir/" <//> "/file" = "./dir/file"@.
+(<//>) :: String -> String -> String
+(<//>) lhs rhs = lhs' ++ "/" ++ rhs'
+  where
+    isSlash = (== '/')
+    lhs' = reverse $ dropWhile isSlash $ reverse lhs
+    rhs' = dropWhile isSlash rhs
+
+-- | Convert (Reader s) to any (MonadState s)
+readerToState
+    :: MonadState s m
+    => Reader s a -> m a
+readerToState = gets . runReader
+
+deriveSafeCopySimple 0 'base ''VerificationRes
+
+-- | A helper for simple error handling in executables
+eitherPanic :: Show a => Text -> Either a b -> b
+eitherPanic msgPrefix = either (error . (msgPrefix <>) . show) identity
+
+-- | This function performs checks at compile-time for different actions.
+-- May slowdown implementation. To disable such checks (especially in benchmarks)
+-- one should compile with: @stack build --flag cardano-sl:-asserts@
+inAssertMode :: Applicative m => m a -> m ()
+#ifdef ASSERTS_ON
+inAssertMode x = x *> pure ()
+#else
+inAssertMode _ = pure ()
+#endif
+{-# INLINE inAssertMode #-}
+
+-- | Remove elements which are in 'b' from 'a'
+diffDoubleMap
+    :: forall k1 k2 v.
+       (Eq k1, Eq k2, Hashable k1, Hashable k2)
+    => HashMap k1 (HashMap k2 v)
+    -> HashMap k1 (HashMap k2 v)
+    -> HashMap k1 (HashMap k2 v)
+diffDoubleMap a b = HM.foldlWithKey' go mempty a
+  where
+    go :: HashMap k1 (HashMap k2 v)
+       -> k1
+       -> HashMap k2 v
+       -> HashMap k1 (HashMap k2 v)
+    go res extKey internalMap =
+        case HM.lookup extKey b of
+            Nothing -> HM.insert extKey internalMap res
+            Just internalMapB ->
+                let diff = internalMap `HM.difference` internalMapB
+                in if null diff
+                       then res
+                       else HM.insert extKey diff res
+
+maybeThrow' :: (Mockable Throw m, Exception e) => e -> Maybe a -> m a
+maybeThrow' e = maybe (throw e) pure
+
+----------------------------------------------------------------------------
+-- NonEmpty
+----------------------------------------------------------------------------
+
+type NE = NonEmpty
+
+neZipWith3 :: (x -> y -> z -> q) -> NonEmpty x -> NonEmpty y -> NonEmpty z -> NonEmpty q
+neZipWith3 f (x :| xs) (y :| ys) (z :| zs) = f x y z :| zipWith3 f xs ys zs
+
+-- | Makes a span on the list, considering tail only. Predicate has
+-- list head as first argument. Used to take non-null prefix that
+-- depends on the first element.
+spanSafe :: (a -> a -> Bool) -> NonEmpty a -> (NonEmpty a, [a])
+spanSafe p (x:|xs) = let (a,b) = span (p x) xs in (x:|a,b)
+
+----------------------------------------------------------------------------
+-- Chronological sequences
+----------------------------------------------------------------------------
+
+newtype NewestFirst f a = NewestFirst {getNewestFirst :: f a}
```

Comment:
:spiral_notepad: Note to self, check these for correctness. It's hard to get these kind of things right.